subsys/bluetooth/mesh/prov.c

```diff
@@ -108,12 +109,16 @@
 #define XACT_NVAL              0xff
 
 enum {
-	REMOTE_PUB_KEY,        /* Remote key has been received */
+	WAIT_PUB_KEY,          /* Waiting for local PubKey to be generated */
 	LINK_ACTIVE,           /* Link has been opened */
+	LINK_ACK_RECVD,        /* Ack for link has been received */
```

Comment:
The link ack is only checked to avoid duplicate callback calls, it's never required for operation. It should probably be treated like the LINK_ACTIVE flag to ensure that we wait for the link ack before we accept any other behavior. 

It'd probably be easiest to change it to `WAIT_LINK_ACK`, and postpone `LINK_ACTIVE` until after the ack is received.

Reply:
The only "problem" is that LINK_ACTIVE is used with `atomic_test_and_set_bit`. If for example the link is already waiting for an ack, ie `WAIT_LINK_ACK` is set, it shouldn't be possible to start another provisioning. That means that any call to start provisioning also has to check against the `WAIT_LINK_ACK` and you then get into a race if it is split between two atomic calls.

One way could be to use the `atomic_cas` maybe. Like in `bt_mesh_pb_gatt_open` for example, `atomic_cas(link.flags, 0, LINK_ACTIVE)` and in `bt_mesh_pb_adv_open`, `atomic_cas(link.flags, 0, PROVISIONER | WAIT_LINK_ACK)`. This would ensure that it is not waiting to ack.

Another option would be to not use `LINK_ACTIVE` as the flag for checking if provisioning is ongoing or not but instead introduce another flag for that.

Any ideas @trond-snekvik?

Reply:
I hadn't thought about that, that's a good point. Sounds like your last suggestion would be the cleanest. Something `LINK_OPEN` to replace `LINK_ACK_RECVD`, but used in both provisioner and provisionee, for instance?

I guess you could also check if either `WAIT_LINK_ACK` or `LINK_ACTIVE` is set when allocating a provisioning instance?

Reply:
> I hadn't thought about that, that's a good point. Sounds like your last suggestion would be the cleanest. Something `LINK_OPEN` to replace `LINK_ACK_RECVD`, but used in both provisioner and provisionee, for instance?

So that would kind of mean renaming `LINK_ACTIVE` to `LINK_OPEN` and then renaming `LINK_ACK_RECVD` to `LINK_ACTIVE`? :thinking: 

> I guess you could also check if either `WAIT_LINK_ACK` or `LINK_ACTIVE` is set when allocating a provisioning instance?

There's not really a way to test for two flags so that could also introduce a race. I think we could use `atomic_cas()` though since the flags should be 0 when starting out.